feat(pagination): experimental pagination style and elements

lovemecomputer opened the pull request:
Closes #1304 

markup & style change to match new experimental design

**Issues:**
- The design of the varied text styling within `<select><option>` elements is not compatible with html spec—`<option>` elements cannot contain children, so we don't have a way of styling part of the text (just the number) in bold without making all of the text bold.
- Because the explanatory text comes *after* the number in the select, a developer using this component will have to write their own javascript to trigger an update of the `<option>` labels when a new selection is made? I don't think this will be good practice to set in a component system—a lot of team will not have the time or eye for detail and end up implementing this incorrectly.

**design spec:**
![screen shot 2018-12-03 at 2 22 21 pm](https://user-images.githubusercontent.com/3846271/49399964-6456a880-f708-11e8-8ff9-550ff2db6351.png)

**live preview to compare:**
https://5c059460ec8a89708177e252--the-carbon-components.netlify.com/component/pagination--default.html

If the select box instead read "Items per page: 40", this would be technically much easier to solve, as the label "Items per page" could be fixed in place. However, the phrasing "40 Items per page" means that this label must flow with the text, and standard HTML elements are not set up to do this.

---

@alisonjoseph @laurenmrice  The design specs also illustrate this design without context:
![screen shot 2018-12-03 at 2 42 32 pm](https://user-images.githubusercontent.com/3846271/49400433-ded3f800-f709-11e8-9d9e-1b5b66d27c19.png)
is this intended as additional/optional functionality, or as the responsive/small-screen state of the pagination component?

lovemecomputer committed:
feat(pagination): experimental pagination style and elements

lovemecomputer committed:
Merge branch 'master' of github.com:IBM/carbon-components into pagination

joshblack commented on src/components/pagination/pagination.hbs:
```diff
@@ -1,57 +1,90 @@
+          data-items-per-page>
+          {{#each itemsPerPageChoices}}
+          <option class="{{@root.prefix}}--select-option" value="{{value}}" {{#if selected}} selected{{/if}}>{{label}}
+            </option> {{/each}} </select> <svg class="{{@root.prefix}}--select__arrow" width="10" height="5" viewBox="0 0 10 5">
```
Can use the `carbon-icon` helper for icons in the updated set (if needed). Docs: https://github.com/IBM/carbon-components/blob/master/docs/experimental.md#faq

lovemecomputer replied:
@joshblack we're only using svg in the v9 version, v10/X is using those helpers! is this proper implementation?

joshblack replied:
My bad!

laurenmrice reviewed (changes requested):
@lovemecomputer 
Hello! 

- The numbered 1,2,3,4,5,6… design is just a secondary alternative to the main component you have created.  :)

- Lets take out all of the semibold and leave the text regular. 'Items' does not need to be capitalized.
- There is a lot of extra padding before and after the ‘1 of 5’ compared to the other sections. Keep the distances 16px/1 rem from the edge of the border line separator. It should also read '1 to 5 pages' and then the caret.
- There are some dotted lines coming up when the focus is activated on the dropdowns that need to go away.
- And I know this is picky but the caret icons look a smidge high in the space compared to the chevrons. If we could bring them down a little that would be great!

aledavila commented:
Hey @laurenmrice '1 to 5 pages' doesn't seem correct. Does that mean its showing 1 to 5 pages (the opposite of pagination) ? 

@lovemecomputer 
- Firefox- Dotted lines on focus. Tabbing not working for dropdowns
- Safari - tabbing not working for arrows
- Chrome - no bugs

aledavila committed:
Merge branch 'master' into pagination

laurenmrice commented:
@aledavila @lovemecomputer  I meant to say it should read '1 of 5 pages' in this example. This can help the user jump to a specific page instead of having to manually use the caret until they get to the page they want. 

<img width="907" alt="screen shot 2018-12-04 at 10 00 58 am" src="https://user-images.githubusercontent.com/43969356/49455154-46428400-f7ac-11e8-97a3-b7471744539d.png">

aledavila commented:
Thanks I see what you mean 'pages' is not on the example sorry 😄

lovemecomputer committed:
fix(pagination): align icons, non-bold text, copy, left-align text

lovemecomputer committed:
Merge branch 'pagination' of github.com:lovemecomputer/carbon-components into pagination

lovemecomputer commented:
@laurenmrice the dotted outline is actually a global issue, good catch! opened a separate issue for it here: xhttps://github.com/IBM/carbon-components/issues/1526

we addressed most of your issues here, please check it out! https://deploy-preview-1518--the-carbon-components.netlify.com/component/pagination--default.html

@aledavila and i will solve the cross-browser issues you pointed out i firefox later on!

laurenmrice reviewed (approved):
Everything looks good to me besides the dotted line issue in focus ! 🎉

asudoh commented on src/components/pagination/_pagination.scss:
```diff
@@ -130,6 +130,10 @@ $css--helpers: true;
     }
 
+    .#{$prefix}--select-input:focus .#{$prefix}--select-option-label {
+      display: none;
```
Is this change here to stable intentional...?

aledavila replied:
No meant to be in experimental.

asudoh reviewed (commented):
Looking nice basically @lovemecomputer - Another question is whether `v2` logic is intentional - Thanks!

aledavila committed:
Merge branch 'master' into pagination

lovemecomputer committed:
fix(pagination): correct CSS rule versions

lovemecomputer committed:
Merge branch 'pagination' of github.com:lovemecomputer/carbon-components into pagination

lovemecomputer committed:
fix(pagination): update CSS rules for cross-browser

(requires _select.css correction)

lovemecomputer commented:
looks like everything is fixed in the latest commit! the icon sizing is a little off—this is actually a global fix that will be merged in with another PR [here](https://github.com/IBM/carbon-components/pull/1537)

asudoh reviewed (commented):
Latest looks good to me, except one thing - We need to keep V1 markup until we release `v10`. Thanks!

lovemecomputer committed:
fix(pagination): display proper variants and add v1 back

lovemecomputer commented:
@asudoh thanks! big shoutout to @joshblack who logged on to a screenshare and vs code sharing session with me and @aledavila to solve this!! thanks y'all

aledavila committed:
Merge branch 'master' into pagination

asudoh commented on src/components/pagination/pagination.config.js:
```diff
@@ -50,29 +52,48 @@ const pageNumberChoices = [
+  },
+]
+  .filter(variant => {
+    if (componentsX) {
```
@lovemecomputer Do you want to try `variant.hidden`?

aledavila committed:
Merge branch 'master' into pagination

jnm2377 commented on src/components/pagination/_pagination.scss:
```diff
@@ -189,3 +189,195 @@ $css--helpers: true;
+    color: $text-02;
+    display: none;
+    padding-right: $spacing-2xs;
+    padding-top: 1px;
```
Was just wondering if it would be better to make this `rem(1px)`? Or the spacing token for `1px`? `$spacing-4xs`.

jnm2377 commented on src/components/pagination/_pagination.scss:
```diff
@@ -189,3 +189,195 @@ $css--helpers: true;
+    cursor: pointer;
+    height: 100%;
+    margin: 0;
+    padding: 0 14px;
```
Again, maybe not required. I just noticed that we use `rem` for the most things, so I was wondering if this should be `rem(14px)` instead?

jnm2377 commented on src/components/pagination/_pagination.scss:
```diff
@@ -189,3 +189,195 @@ $css--helpers: true;
+
+  .#{$prefix}--pagination__button--no-index {
+    border-right: 0;
+    margin-right: 1px;
```
Same as before

jnm2377 commented on src/components/pagination/_pagination.scss:
```diff
@@ -189,3 +189,195 @@ $css--helpers: true;
+  }
+
+  .#{$prefix}--pagination--inline {
+    height: 42px;
```
same as before

jnm2377 reviewed (approved):
Other than those questions I had, it looks great!

lovemecomputer commented:
@jnm2377 yeah that makes sense, def agree we should be using spacing tokens here! will update

lovemecomputer committed:
fix(pagination): change px rules to tokens and rem()

lovemecomputer committed:
Merge branch 'master' of github.com:IBM/carbon-components into pagination

lovemecomputer committed:
Merge branch 'pagination' of github.com:lovemecomputer/carbon-components into pagination

lovemecomputer commented:
@jnm2377 thank u for the analysis! updated and pushed.
